spec/message.json

```diff
@@ -0,0 +1,162 @@
+        "type": { "const": "variable" },
+        "name": { "type": "string" }
+      },
+      "required": ["type", "name"]
+    },
+    "value": {
+      "oneOf": [{ "$ref": "#/$defs/literal" }, { "$ref": "#/$defs/variable" }]
+    },
+
+    "function": {
+      "type": "object",
+      "properties": {
+        "type": { "const": "function" },
+        "kind": { "enum": ["open", "close", "value"] },
+        "name": { "type": "string" },
+        "operand": { "$ref": "#/$defs/value" },
+        "options": {
+          "type": "array",
+          "items": {
+            "type": "object",
+            "properties": {
+              "name": { "type": "string" },
+              "value": { "$ref": "#/$defs/value" }
+            },
+            "required": ["name", "value"]
+          }
+        }
+      },
+      "required": ["type", "kind", "name"]
+    },
+    "reserved": {
+      "type": "object",
+      "properties": {
+        "type": { "const": "reserved" },
+        "sigil": {
+          "enum": ["!", "@", "#", "%", "^", "&", "*", "<", ">", "?", "~"]
+        },
+        "source": { "type": "string" },
+        "operand": { "$ref": "#/$defs/value" }
+      },
+      "required": ["type", "sigil", "source"]
+    },
+
+    "text": {
+      "type": "object",
+      "properties": {
+        "type": { "const": "text" },
+        "value": { "type": "string" }
+      },
+      "required": ["type", "value"]
+    },
+    "expression": {
+      "type": "object",
+      "properties": {
+        "type": { "const": "expression" },
+        "body": {
+          "oneOf": [
+            { "$ref": "#/$defs/literal" },
+            { "$ref": "#/$defs/variable" },
+            { "$ref": "#/$defs/function" },
+            { "$ref": "#/$defs/reserved" }
+          ]
+        }
+      },
+      "required": ["type", "body"]
+    },
+    "pattern": {
+      "type": "object",
+      "properties": {
+        "body": {
+          "type": "array",
+          "items": {
+            "oneOf": [
+              { "$ref": "#/$defs/text" },
+              { "$ref": "#/$defs/expression" }
+            ]
+          }
+        }
+      },
+      "required": ["body"]
+    },
+
+    "declarations": {
+      "type": "array",
+      "items": {
+        "type": "object",
+        "properties": {
+          "target": { "$ref": "#/$defs/variable" },
+          "value": { "$ref": "#/$defs/expression" }
+        },
+        "required": ["target", "value"]
+      }
+    },
+    "variant-key": {
+      "oneOf": [
+        { "$ref": "#/$defs/literal" },
+        {
+          "type": "object",
+          "properties": {
+            "type": { "const": "*" },
+            "value": { "type": "string" }
+          },
+          "required": ["type"]
+        }
+      ]
+    },
+    "message": {
+      "type": "object",
+      "properties": {
+        "type": { "const": "message" },
+        "declarations": { "$ref": "#/$defs/declarations" },
+        "pattern": { "$ref": "#/$defs/pattern" }
```

Comment:
Seems like it might be simpler to have a single `message`, which has a `body` attribute that's one of `pattern` or `select`, and then `select` would just have the `selectors` and `variants` fields? (In other words, `declarations` would be factored out.)

Reply:
There's also the approach @stasm has suggested, i.e. expressing single-pattern messages as having an empty list for selectors, and a single entry in variants with a corrspondingly empty list of keys.

As I mentioned on the call, I'd much prefer landing something first and then iterating on it, rather than sorting out the details all in one go.

Reply:
I filed #437 to continue discussing this.